compiler/packages/babel-plugin-react-compiler/src/Entrypoint/Program.ts

```diff
@@ -541,30 +540,26 @@ export function compileProgram(
   if (moduleScopeOptOutDirectives.length > 0) {
     return;
   }
-
+  let gating: null | {
```

Comment:
Changes in this file are all code movement.
- tryParseExternalFunction is moved to `parsePluginOptions`, which runs before this function
- collect `referencedBeforeDeclared` set here instead eagerly bailing out. This is to prepare for the next PR https://github.com/facebook/react/pull/32598. (Also happy to merge these into a single changeset if that's easier to review)